[ML] Improve Explorer Chart labels.

walterra opened the pull request:
Improves the display of the Explorer Chart labels to fix the following issues:

- Long chart labels could be cut off, so it's not possible to tell what entity fields a chart is referring to. A workaround is to hover the info icon tooltip but that's really slow and cumbersome if you have to do it for every chart.

![image](https://user-images.githubusercontent.com/230104/46024220-9afbbb80-c0e6-11e8-9de5-792102000f41.png)

- The list of entity fields and its values is an unformatted text blob which makes it hard to read and tell which values refer to which field

![image](https://user-images.githubusercontent.com/230104/46024419-0c3b6e80-c0e7-11e8-9d98-35b2d21128e2.png)

Proposed solution:

- If any of the chart labels is longer than 60 chars, the entity fields will wrap to a new line (for all charts to a achieve a consistent look).
- Entity fields use `EuiBadge` and some custom formatting to make it easier to see field/value pairs.

![image](https://user-images.githubusercontent.com/230104/46024633-866bf300-c0e7-11e8-91bc-fa1b2b07a557.png)

- If the detector description is too long, it still uses `ellipsis` for `text-overflow`:

![image](https://user-images.githubusercontent.com/230104/46143227-a28fa180-c259-11e8-9610-211755faedc7.png)

- If the entity badges are too long, they will be just cut off to the right. There's no simple CSS fix for that, we cannot use `ellipsis` and we don't want to wrap those badges again because then multiple charts could have different heights. I experimented with gradients but that turned out to be somewhat unreliable. I still consider this a good enough improvement compare to the previous version and would like to leave a tweak for that to a follow up PR. 

![image](https://user-images.githubusercontent.com/230104/46081999-39972380-c19e-11e8-8eaa-4428e54ef0cc.png)

- If there are mixed detectors with and without entity fields and the existing one wrap, multiple charts are aligned considered the height of the entity fields on display:

![image](https://user-images.githubusercontent.com/230104/46137451-f1353f80-c249-11e8-9e37-a2267ea5ee90.png)

- Additionally, this changes the link to the single series viewer from custom code using a Font Awesome icon to use `EuiButtonEmpty` with the same EUI based icon and a tooltip.

![image](https://user-images.githubusercontent.com/230104/46081054-d0aeac00-c19b-11e8-889e-fe58e21920e0.png)

Note: The threshold of 60 chars was chosen after testing the responsive layout and how we determine how many charts to show per row. I tried to do this without completely revamping/rethinking how this chart header is done, but after migrating the view-link to EUI and adding tests, well, here's what we got :). How this header is done could be further improved but I'd consider this a good enough solution for now. The main reason I did this is to make the entityFields information more readable and not being hidden easily. Let me know what you think.

----

Part of #21163.

elasticmachine commented:
Pinging @elastic/ml-ui

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3428/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3432/)

jgowdyelastic commented on x-pack/plugins/ml/public/util/chart_utils.js:
```diff
@@ -228,6 +228,15 @@ export function getTickValues(startTimeMs, tickInterval, earliest, latest) {
+
+// Checks if the string length of a chart label (detector description
+// and entity fields) is above LABEL_WRAP_THRESHOLD.
+export function isLabelLengthAboveThreshold({ detectorLabel, entityFields }) {
```
this function is using character counting to measure whether it should wrap. i don’t think the font is monospaced, which will result in inconsistent results.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3444/)

alvarezmelissa87 commented on x-pack/plugins/ml/public/explorer/explorer_charts/components/explorer_chart_label/explorer_chart_label.js:
```diff
@@ -0,0 +1,69 @@
+  const entityFieldBadges = entityFields.map((entity, j) => {
+    return <span key={j}><ExplorerChartLabelBadge entity={entity} /> </span>;
+  });
+
```
Just a quick note - we generally want to avoid using indexes as keys and instead use a unique identifier.

Since React uses the key to identify DOM elements and see if they've changed. So if the order changes for some reason there could be bugs due to React thinking that because the key is the same, the content is the same. (more info here if you're curious: https://medium.com/@vraa/why-using-an-index-as-key-in-react-is-probably-a-bad-idea-7543de68b17c)

I haven't dug as deeply for fragments regarding this but I'd err on the side of caution.

walterra replied:
Nice find, I built a custom key for the outer `.map()` part in the same file but let hat slip through here :)

walterra replied:
@alvarezmelissa87 pushed an update to use a proper key based on the `entity` data.

alvarezmelissa87 commented on x-pack/plugins/ml/public/explorer/explorer_charts/explorer_charts_container.js:
```diff
@@ -7,18 +7,31 @@
+import { ExplorerChartLabel } from './components/explorer_chart_label';
+
+const textTooManyBuckets = 'This selection contains too many buckets to be displayed.' +
+  'The dashboard is best viewed over a shorter time range.';
```
Bit of a nitpick... Could probably use template literals here to keep in line with the style guide.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3517/)

walterra commented:
@peteharverson I pushed some updates addressing your questions:

1. If there are mixed detectors with and without entity fields and the existing one wrap, multiple charts are aligned considered the height of the entity fields on display. I updated the PR description with a screenshot. The charts without entity fields show a gap so the chart itself is on the same height as others and the overall container height stays the same for all charts.
2. The missing info icon is now also fixed, changed the structure so the `ellipsis` overflow is only applied to the text itself.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3522/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3529/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3541/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3722/)
